server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -853,15 +858,20 @@ public IndexShard createShard(
         RecoveryState recoveryState = indexService.createRecoveryState(shardRouting, targetNode, sourceNode);
         IndexShard indexShard = indexService.createShard(shardRouting, globalCheckpointSyncer, retentionLeaseSyncer);
         indexShard.addShardFailureCallback(onShardFailure);
-        indexShard.startRecovery(recoveryState, recoveryTargetService, recoveryListener, repositoriesService, mapping -> {
+        indexShard.startRecovery(recoveryState, recoveryTargetService, recoveryListener, repositoriesService, (mapping, listener) -> {
             assert recoveryState.getRecoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS
                 : "mapping update consumer only required by local shards recovery";
-            client.admin()
-                .indices()
-                .preparePutMapping()
-                .setConcreteIndex(shardRouting.index()) // concrete index - no name clash, it uses uuid
-                .setSource(mapping.source().string(), XContentType.JSON)
-                .get();
+            client.execute(
+                clusterService.state().nodes().getMinNodeVersion().onOrAfter(Version.V_8_8_0)
+                    ? AutoPutMappingAction.INSTANCE
+                    : PutMappingAction.INSTANCE,
+                new PutMappingRequest().setConcreteIndex(shardRouting.index())
+                    .setConcreteIndex(shardRouting.index()) // concrete index - no name clash, it uses uuid
+                    .source(mapping.source().string(), XContentType.JSON)
+                    .timeout(TimeValue.MAX_VALUE)
+                    .masterNodeTimeout(TimeValue.MAX_VALUE),
+                new ThreadedActionListener<>(threadPool.generic(), listener.map(ignored -> null))
```

Comment:
This slightly changes behavior during shutdown I think in that, we'd never complete the listener. I doubt that we wait for recoveries to complete anywhere in our shutdown sequence though, but wanted to be sure you had this on your radar too.

Reply:
++ yes we fork to `GENERIC` all over the place during recovery already, see `org.elasticsearch.indices.recovery.RemoteRecoveryTargetHandler#executeRetryableAction`.